### PR TITLE
[Bug] fix 1.23k formatting

### PR DESCRIPTION
**Pull request by b2kdaman**
Added custom formatter support

Signed-off-by: Valentine <vpanchin@gmail.com>

### CODE REVIEW

**Comment by b2kdaman** [comment]
@heshan0131 @macrigiuseppe

**Comment by heshan0131** [comment]
Can you explain what this bug is about? why do you need a custom formatted on a default format?

**Comment by b2kdaman** [comment]
> Can you explain what this bug is about? why do you need a custom formater on a default format?

It's for easier values representation. Basially limits value to 3 digits, so we can see values <1000

**Comment by macrigiuseppe** [comment]
> > Can you explain what this bug is about? why do you need a custom formater on a default format?
> 
> It's for easier values representation. Basially limits value to 3 digits, so we can see values <1000

This is too specific for your use case and it may not work with all the other scenarios. I personally don't think we should implement such a specific formatter

**Comment by thunter009** [comment]
> > > Can you explain what this bug is about? why do you need a custom formater on a default format?
> > 
> > 
> > It's for easier values representation. Basially limits value to 3 digits, so we can see values <1000
> 
> This is too specific for your use case and it may not work with all the other scenarios. I personally don't think we should implement such a specific formatter

This format string is useful for scenarios where you have data that is power law distributed and lots of values fall <1000 but a few values are > 1000.

**Comment by heshan0131** [comment]
with the current formatter `.3s`, it formats numbers as decimal notation with an SI prefix, rounded to 3 significant digits.

So expected output should be
```
123 -> 123
1223 -> 1.22k
```

You can perhaps add a `~` to `.3~s` so it trims the 0s

```
9000 -> 9k (instead of 9.00k)
```

Is this something work for you?

File: test/browser/index.js [hunk 1]
```diff
@@ -28,6 +28,9 @@ require('./components');
 // test layers
 require('./layer-tests');
 
+// test formatters
+require('./formatters');
```
Comment (heshan0131) [inline test/browser/index.js:32]: if you only have 1 formatter test, just directly import here

`require('./formatters-tests')`

**Comment by thunter009** [comment]
> with the current formatter `.3s`, it formats numbers as decimal notation with an SI prefix, rounded to 3 significant digits.
> 
> So expected output should be
> 
> ```
> 123 -> 123
> 1223 -> 1.22k
> ```
> 
> You can perhaps add a `~` to `.3~s` so it trims the 0s
> 
> ```
> 9000 -> 9k (instead of 9.00k)
> ```
> 
> Is this something work for you?

This would work perfectly.

Ideally, I'd love to have support for the user entering/editing arbitrary d3 format strings as that feature would obviate the need for building this specific format option. This was a feature in Superset that I absolutely **loved**

File: test/browser/index.js (hunk 1 above)
Comment (b2kdaman) [reply test/browser/index.js:32]: But what if we would have additional formatters in the future?

**Comment by b2kdaman** [comment]
> with the current formatter `.3s`, it formats numbers as decimal notation with an SI prefix, rounded to 3 significant digits.
> 
> So expected output should be
> 
> ```
> 123 -> 123
> 1223 -> 1.22k
> ```
> 
> You can perhaps add a `~` to `.3~s` so it trims the 0s
> 
> ```
> 9000 -> 9k (instead of 9.00k)
> ```
> 
> Is this something work for you?

It actually doesn't
image
 
I was playing with d3 formatter and it has some issues. This is why I had to create this additional "formatter"

**Comment by b2kdaman** [comment]
@heshan0131 sorry for not being clear from the start. This is an old fix tho. 
So the whole problem was that when there are only two digits supplied we would see: `99 -> 99.0`
So this relsolves issue with numbers `< 100`

**Comment by heshan0131** [comment]
depends on which version of `d3-format` you are using. the result of `~s` will different

**Comment by b2kdaman** [comment]
> depends on which version of `d3-format` you are using. the result of `~s` will different

Have you implemented this in kepler/unfolded already?

**Pull request closed without merging**